### PR TITLE
avoid stackoverflow, when accessing entities with lazy loaded references via JSON-Request

### PR DESCRIPTION
**Pull request by Gab0rB**
Here is a solution to avoid stackoverflow errors, when accessing entities with lazy loaded references via JSON-Request. 
I got this solution from this thread: http://stackoverflow.com/questions/30689697/jackson-hibernate-lots-of-problems

steps, which has to be done:
1. define a bean which extends **OpenEntityManagerInViewFilter**
2. define interfaces for the views, which should be shown in the frontend 
(as example i created a View for **Project**)
3. annotate the columns in the entity-objects, which should be shown by the defined interface
(**@JsonView(Views.Project.class)**)
4. create a Controller, which overrides the standard-CrudRepository method(**in this case findAll**) by mapping it to the same URL (**/projects**) and annotate it with **@JsonView(Views.Project.class)**

comment:
- the transaction-context of ProjectRepository changed by adding the OpenEntityManagerInViewFilter, so the PlantPageController methods had to be annotated with @Transactional to work correctly, else you get lazy loading exceptions, when trying to acces this lazy loaded objects(in this case getProjectArticles)

Please have a look at this and share your opinion

### CODE REVIEW

File: user/src/main/java/org/dicadeveloper/weplantaforest/projects/Price.java [hunk 1]
```diff
@@ -37,6 +41,7 @@
     private ScontoType scontoType = ScontoType.NONE;
 
     @Column(name = "_amount", precision = 7, scale = 2)
+    @JsonView({Views.Project.class})
```
Comment (mschaaf) [inline user/src/main/java/org/dicadeveloper/weplantaforest/projects/Price.java]: Does this mean that the view project contains also an price?
Where is this view defined?

Comment (Gab0rB) [inline user/src/main/java/org/dicadeveloper/weplantaforest/projects/Price.java]: This means, that the view project, contains articles, which contains a price with an amount.
It's the same structure like the entities/tables are defined.
With this View interface you can only define which elements(columns/references) should be visible when calling the url, you can't change the structure of the entity or their references to other entities.

You define the View by adding/removing  @JsonView({Views.Project.class}) to the elements you want to show with this View, so there is no explicit definition of it.

Comment (Gab0rB) [inline user/src/main/java/org/dicadeveloper/weplantaforest/projects/Price.java]: her's an example for the first object when calling /projects:
sorry for the bad formatting, it seems, that whitespaces won't be recognized here

 "0": {
    "name": "Project 0 von Johannes",
    "description": "dksgny.d, mdfgnmn snfad,ng ,ydfng. ,ydfgnk.<sngdk< sglkbsnglkdfnksghnl<k njdjg nsgyö< ögn kl< bsflkjsb gkjs kgs ns< lödgksndlkgnöd<kl dykdyn ökd ökshö<g dysh ögskgös Hskg khoglksg",
    "longitude": 13.607287,
    "latitude": 52.462654,
    "manager": {
      "name": "Johannes"
    },
    "articles": [
      {
        "articleId": 1,
        "treeType": {
          "name": "Buche",
          "annualCo2SavingInTons": 0.02
        },
        "price": {
          "amount": 2.26
        },
        "amount": 9
      },
      {
        "articleId": 2,
        "treeType": {
          "name": "Kiefer",
          "annualCo2SavingInTons": 0.02
        },
        "price": {
          "amount": 5.32
        },
        "amount": 483
      },
      {
        "articleId": 3,
        "treeType": {
          "name": "Birke",
          "annualCo2SavingInTons": 0.02
        },
        "price": {
          "amount": 1.35
        },
        "amount": 412
      }
   ]
  }

File: user/src/main/java/org/dicadeveloper/weplantaforest/Views.java [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package org.dicadeveloper.weplantaforest;
+
+public interface Views {
+    public static interface Public {}
+    public static interface Project extends Public{}
```
Comment (mschaaf) [inline user/src/main/java/org/dicadeveloper/weplantaforest/Views.java]: Ah, das ist die Definition. Das kann mir gefallen. Gibt es nur die Möglichkeit die Properties mit der JSonView Annotation zu markieren oder kann man das auch direkt hier machen? Damit man das alles schön im Überblick hat an einer Stelle.

File: user/src/main/java/org/dicadeveloper/weplantaforest/projects/Price.java (hunk 1 above)
Comment (mschaaf) [inline user/src/main/java/org/dicadeveloper/weplantaforest/projects/Price.java]: I update your comment to apply formatting. github allows to put preformatted text in 

\```

to achieve this.

File: user/src/main/java/org/dicadeveloper/weplantaforest/Views.java (hunk 2 above)
Comment (Gab0rB) [inline user/src/main/java/org/dicadeveloper/weplantaforest/Views.java]: Das hätte was ja, allerdings habe ich bisher keine Möglichkeit gefunden.

Comment (mschaaf) [inline user/src/main/java/org/dicadeveloper/weplantaforest/Views.java]: Wäre das i Tüpfelchen ist aber so auch ganz schön.

**Comment by mschaaf** [comment]
We decided to go that way, so please feel free to merge and close.

**Comment by Gab0rB** [comment]
Merging isn't necessary here. We don't use /projects atm in the frontend, so the defined interface is more an example than a usecase for the application.
I will implement this, if we really need a query to one of the implemented entities with lazy loaded references in the frontend.

**Pull request closed without merging**